opesci/fields.py

```diff
@@ -201,51 +202,70 @@ def set(self, dimension, direction):
         staggered[direction] = True
         Field.set(self, dimension, staggered)
 
-    def set_free_surface(self, d, b, side):
+    def set_free_surface(self, d, b, side, algo=0):
         """
         - set free surface boundary condition to boundary d, at index b
         :param d: direction of the boundary surface normal
         :param b: location of the boundary (index)
         :param side: lower boundary (0) or upper boundary (1)
+        :param algo: which algorithm to use to compute ghost cells
         - e.g. set_free_surface([t,x,y,z],1,2,0)
         set y-z plane at x=2 to be lower free surface
         - ghost cells are calculated using reflection of stress fields
         - store the symbolic equations to populate ghost cells in self.bc
         """
-        # use this stress field to solve for ghost cell expression
-        field = self.sfields[d]
-        expr = field.dt
-        idx = list(self.indices)
-        # create substituion dictionary
-        dict1 = {}
-        derivatives = get_all_objects(expr, DDerivative)
-        for deriv in derivatives:
-            # using 2nd order approximation
-            dict1[deriv] = deriv.fd[1]
-        expr = expr.subs(dict1)
-        if self.staggered[d]:
-            # if staggered, solve ghost cell using T'[b]=0 (e.g. W at z surface, using Tzz)
-            eq = Eq(expr)
-            shift = hf
-            t = b - hf  # real boundary location
-        else:
-            # if not staggered, solve ghost cell using T'[b-1/2]=T'[b+1/2] (e.g. U at z surface, using Txz)
-            eq = Eq(expr.subs(idx[d], idx[d]-hf),
-                    expr.subs(idx[d], idx[d]+hf))
-            shift = 1
-            t = b
-
-        idx[d] -= ((-1)**side)*shift
-        lhs = self[idx]
-        rhs = solve(eq, lhs)[0]
-        lhs = lhs.subs(self.indices[d], t)
-        rhs = self.align(rhs.subs(self.indices[d], t))
-
-        # change ti to t+1
-        lhs = lhs.subs(idx[0], idx[0]+1)
-        rhs = rhs.subs(idx[0], idx[0]+1)
-
-        self.bc[d][side] = [Eq(lhs, rhs)]
+        if algo == 0:
```

Comment:
Also, can we use string names for the algorithms rather than some magic numbers.